src/guards/slice_memory_guard.rs

```diff
@@ -50,6 +51,53 @@ impl<'a, T> SliceMemoryGuard<'a, T> {
         }
         SliceMemoryGuard { memory }
     }
+
+    /// Initialize memory guard using given iterator.
+    /// Automatically shrink's memory to given items' count.
+    /// `Ok(guard)` will be returned in this case.
+    ///
+    /// If items' count is too large to place in memory, moves it into new `Vec` and continue collecting into it.
+    /// `Err(vec)` will be returned in this case.
+    #[inline]
+    pub unsafe fn new_from_iter(memory: &'a mut [MaybeUninit<T>], mut iter: impl Iterator<Item=T>) -> Result<Self, Vec<T>> {
+        // Fulfilling placed memory
+        for (index, item) in memory.into_iter().enumerate() {
+            match iter.next() {
+                // While iterator returns new value, write it
+                Some(value) => {
+                    write(item.as_mut_ptr(), value);
+                }
+                // When it returns None then slicing memory and returning the guard
+                None => {
+                    return Ok(SliceMemoryGuard {
+                        memory: &mut memory[0..index],
+                    });
+                }
+            }
+        }
+
+        if let Some(next_item) = iter.next() {
+            let memory = transmute::<&'a mut [MaybeUninit<T>], &'a mut [T]>(memory);
+            // If iterator still contains values to return, collect it into the vector
+            let mut vec = Vec::<T>::with_capacity(
+                // We cannot trust the `size_hint` anymore
+                memory.len() + 1
+            );
+
+            // First, copying already fulfilled memory into the heap
+            vec.set_len(memory.len());
+            copy_nonoverlapping(memory.as_mut_ptr(), vec.as_mut_ptr(), memory.len());
```

Comment:
do we actually need that transmute higher up? Would it be better to just do `memory.as_mut_ptr() as *mut T` here instead?

Reply:
Actually, no. I just forget to delete this line.